### PR TITLE
WIP try to document all the different types of token in the FxA ecosystem

### PR DESCRIPTION
**Pull request by rfk**
Opening this as an early draft for visibility - I wanted a reference on all the different types of token in the FxA ecosystem and what we use them for, so I started adding one to the ecosystem docs. All feedback welcome even at this early stage, especially if you have suggestions for things you'd like to see in this document!

[Rendered View](https://github.com/mozilla/ecosystem-platform/blob/fxa-token-types/docs/fxa-engineering/fxa-token-types.md)

/cc @shane-tomlinson @vbudhram

### CODE REVIEW

File: docs/fxa-engineering/fxa-token-types.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+Session tokens are an ***FxA Hawk Token*** where the 32-byte string is generated by the server. They can be obtained by:
+* The successful creation of a new account, via the `/account/create` endpoint.
+* A successful login to a new account, via the `/account/login` endpoint.
+* Duplicating an existing `sessionToken`, via the `/session/duplicate` endpoint.
+* Providing the id of a previous `sessionToken` when changing the account password, via the `/password/change/finish`
+  endpoint.
+* Requesting the `https://identity.mozilla.com/tokens/session` scope during an OAuth flow.
+
+Session tokens are intended only for authenticating communication with the FxA servers themselves, so RP servers
```
Comment (shane-tomlinson) [inline docs/fxa-engineering/fxa-token-types.md]: `authenticating` => `authenticated`

File: docs/fxa-engineering/fxa-token-types.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+* The types of authentication performed by the user, and level of confidence in their identity.
+* Timestamp at which authentication last occurred (which can be increased by prompting the user to re-enter their
+  password).
+* Timestamp at which the token was last used (truncated to a few hours granularity to reduce db write load).
+* Details of the user-agent string from the last use of the token.
+
+This information may be used by the FxA server to make authorization decisions, and RPs may request that a user
+be authenticated to a certain level of assurance (such as requiring that the sessionToken be last-authenticated
+within a certain amount of time, or authenticated via 2FA).
```
Comment (shane-tomlinson) [inline docs/fxa-engineering/fxa-token-types.md]: We don't have `within a certain amount of time` right now, do we?

File: docs/fxa-engineering/fxa-token-types.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+## Verified Sessions
+
+As an additional security measure against [credential-stuffing
+attacks](https://blog.mozilla.org/services/2016/04/09/stolen-passwords-used-to-break-into-firefox-accounts/),
+FxA has the notion of a "verified session", which is intended to indicate a higher level of confidence in the
+authenticity of the user.
+
+A session is considered verified if the user did some additional authentication step *in addition* to proving
```
Comment (shane-tomlinson) [inline docs/fxa-engineering/fxa-token-types.md]: This feels very AALy to me, is there room for a model refactor here?

File: docs/fxa-engineering/fxa-token-types.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+Sessions may also be considered verified based on server-side heuristics, such as whether the account is newly-created
+or whether we've seen previous verified logins form the same IP address.
+
+Certain actions on the account can only be performed with a verified session, including:
+* Modifying email or 2FA settings
+* Authorizing OAuth grants that involve key-bearing scopes
+
+## Multi-Factor Auth
+
+If the user enalbes two-step authentication on their account, then each sessionToken will keep track of
```
Comment (shane-tomlinson) [inline docs/fxa-engineering/fxa-token-types.md]: `enalbes` => `enables`

File: docs/fxa-engineering/fxa-token-types.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+They're for signin. Link to OIDC spec, and explain why they're useful.
+
+# KeyFetch Tokens
+
+They're for fetching keys. RPs shouldn't need to now this, because they use the scoped-keys flow.
+Talk a bit about verification and how it relates to sessionToken verification.
+Explain the whole scrypt-on-the-server thing, and the way we encrypt the keys for communicating
+back to the client.
+
+# Password Change Tokens
```
Comment (shane-tomlinson) [inline docs/fxa-engineering/fxa-token-types.md]: These seem very much like OAuth access codes with a specific scope, e.g., "password-change", "password-reset", "account-reset".

File: docs/fxa-engineering/fxa-token-types.md (hunk 3 above)
Comment (rfk) [reply docs/fxa-engineering/fxa-token-types.md]: Oh my, there is *so much* room for a model refactor here 😆

**Comment by clouserw** [comment]
I rebased this off HEAD and updated some of the wording.  It's not nearly as complete as rfk would have left it, of course, but as Shane said, it's better than what we have.   @mozilla/fxa-devs for review

File: docs/reference/token-types.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+# Token Usage Mechanisms
+
+A ***Bearer Token*** is an opaque string that can be used by any client in posession of the token, simply by presenting
+it verbatim in the request to the server (e.g. in the `Authorization` header). These are the simplest types of token
+for clients to use, but allow for interception and replay attacks. In particular, the server that receives a Bearer Token
+is capable of extracting it and using it to perform other API calls.
+
+A ***Hawk Token*** consists of an opaque string id and corresponding secret key. To use the token, clients must use the
+secret key to sign outgoing requests using [the scheme described here](https://hapi.dev/family/hawk/). This is more
```
Comment (chenba) [inline docs/reference/token-types.md:43]: Hawk should be linked to https://github.com/mozilla/hawk/blob/main/API.md or https://github.com/mozilla/hawk

File: docs/reference/token-types.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+# Session Metadata
+
+Since session tokens represent a signed-in user, the server maintains a lot of additional metadata about the user's
+signed-in state. Properties of the session token include:
+
+* The types of authentication performed by the user, and level of confidence in their identity.
+* Timestamp at which authentication last occurred (which can be increased by prompting the user to re-enter their
+  password).
+* Timestamp at which the token was last used (truncated to a few hours granularity to reduce db write load).
+* Details of the user-agent string from the last use of the token.
```
Comment (chenba) [inline docs/reference/token-types.md:81]: The user's ip based geo location is also on the session token.

Comment (clouserw) [reply docs/reference/token-types.md:81]: added to new version

**Review by chenba — APPROVED** [review]

**Pull request merged**